src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs

```diff
@@ -206,6 +206,14 @@ void CheckAndWarnOnReflectionAccess(in MessageOrigin origin, TypeSystemEntity en
                 }
             }
 
+            var diagnosticContext = new DiagnosticContext(
+                origin,
+                _logger.ShouldSuppressAnalysisWarningsForRequires(origin.MemberDefinition, DiagnosticUtilities.RequiresUnreferencedCodeAttribute),
+                _logger.ShouldSuppressAnalysisWarningsForRequires(origin.MemberDefinition, DiagnosticUtilities.RequiresDynamicCodeAttribute),
+                _logger.ShouldSuppressAnalysisWarningsForRequires(origin.MemberDefinition, DiagnosticUtilities.RequiresAssemblyFilesAttribute),
+                _logger);
```

Comment:
This is relatively expensive. I was thinking about this before and I think a better approach would be to split the `CheckAndReport` method below - first check if there's any of the interesting attributes on the target method, if NOT, do nothing else. Otherwise construct the diag context and report it. The point is that constructing the diag context means we have to look for the attributes on the origin, but if there's no warning to report then it's wasted effort.
If there is a warning we will have to look both on the origin and on the called method, but we should only do that when necessary.